File: scripts/validate_eval.js
const fs = require('fs');
const yaml = require('js-yaml');

const dir = 'docs/design/contracts/webui-and-pm-workspace';
const contractPath = `${dir}/ui-contract.md`;

let results = {
  containerValidity: {},
  prdCoverage: {},
  componentTraceability: {},
  logicCoverage: {},
  derivedArtifacts: {},
  conflictsPreview: {}
};

// 1. Container Validity
const content = fs.readFileSync(contractPath, 'utf8');
const yamlMatches = content.match(/^```yaml\n([\s\S]*?)^```/m);
const mermaidMatches = content.match(/^```mermaid\n([\s\S]*?)^```/m);

results.containerValidity = {
  yamlBlockCount: content.split(/^```yaml/m).length - 1,
  mermaidBlockCount: content.split(/^```mermaid/m).length - 1,
  validYaml: false
};

let yamlData = null;
if (yamlMatches && yamlMatches[1]) {
  try {
    yamlData = yaml.load(yamlMatches[1]);
    results.containerValidity.validYaml = true;
  } catch (e) {
    results.containerValidity.error = e.message;
  }
}

// 2. PRD Coverage & 3. Component Traceability & 4. Logic Coverage
let dsIds = [];
let duplicates = [];
let actions = new Set();
let prdCoverage = 0;

if (yamlData) {
  const screens = yamlData.screens || [];
  const routes = yamlData.routes || [];
  const satisfiesCount = (yamlData.metadata && yamlData.metadata.satisfies) ? yamlData.metadata.satisfies.length : 0;
  
  const collectFromElement = (el) => {
    if (el.ds_id) {
      if (dsIds.includes(el.ds_id)) duplicates.push(el.ds_id);
      dsIds.push(el.ds_id);
    }
    if (el.actions) {
      el.actions.forEach(a => actions.add(a.id || a.action_id || a.name || a));
    }
    if (el.children) {
      el.children.forEach(collectFromElement);
    }
    if (el.states) {
      Object.values(el.states).forEach(st => {
        if (st.actions) st.actions.forEach(a => actions.add(a.id || a.action_id || a.name || a));
        if (st.children) st.children.forEach(collectFromElement);
      });
    }
  };
  
  screens.forEach(collectFromElement);
  
  results.prdCoverage = {
    satisfiesCount,
    screensCount: screens.length,
    routesCount: routes.length
  };
  
  results.componentTraceability = {
    uniqueDsIdsCount: new Set(dsIds).size,
    duplicates: duplicates
  };
  
  // Mermaid transitions
  const mermaidLines = mermaidMatches ? mermaidMatches[1].split('\n') : [];
  let mermaidEvents = new Set();
  mermaidLines.forEach(line => {
    const match = line.match(/-->.*?:\s*(.*)/);
    if (match) mermaidEvents.add(match[1].trim());
  });
  
  let unmatchedActions = [];
  actions.forEach(a => {
    if (!mermaidEvents.has(a)) unmatchedActions.push(a);
  });
  
  let unmatchedEvents = [];
  mermaidEvents.forEach(e => {
    if (!actions.has(e)) unmatchedEvents.push(e);
  });
  
  results.logicCoverage = {
    actionsCount: actions.size,
    eventsCount: mermaidEvents.size,
    unmatchedActions,
    unmatchedEvents
  };
}

// 5. Derived Artifacts Consistency
const checkJson = (filename) => {
  const p = `${dir}/${filename}`;
  if (!fs.existsSync(p)) return { exists: false };
  try {
    const data = JSON.parse(fs.readFileSync(p, 'utf8'));
    return { exists: true, valid: true, count: Array.isArray(data) ? data.length : Object.keys(data).length };
  } catch (e) {
    return { exists: true, valid: false };
  }
};

results.derivedArtifacts = {
  storyboards: checkJson('storyboards.json'),
  componentMap: checkJson('component-map.json'),
  layoutRules: checkJson('layout-rules.json'),
  artifactIndex: checkJson('artifact-index.json')
};

results.derivedArtifacts.contextSlicesExists = fs.existsSync(`${dir}/context-slices`);
results.derivedArtifacts.reviewDiagramsExists = fs.existsSync(`${dir}/review-diagrams.md`);

// 6. Conflict & Preview Readiness
results.conflictsPreview = {
  previewManifestExists: fs.existsSync(`${dir}/preview/preview-manifest.json`),
  conflictsExists: fs.existsSync(`${dir}/prd-ds-conflicts.md`)
};

console.log(JSON.stringify(results, null, 2));

File: scripts/validate_eval.py
import yaml
import json
import re
import os

dir_path = 'docs/design/contracts/webui-and-pm-workspace'
contract_path = f'{dir_path}/ui-contract.md'

results = {
    'containerValidity': {},
    'prdCoverage': {},
    'componentTraceability': {},
    'logicCoverage': {},
    'derivedArtifacts': {},
    'conflictsPreview': {}
}

with open(contract_path, 'r', encoding='utf-8') as f:
    content = f.read()

yaml_blocks = re.findall(r'^```yaml\n([\s\S]*?)^```', content, re.MULTILINE)
mermaid_blocks = re.findall(r'^```mermaid\n([\s\S]*?)^```', content, re.MULTILINE)

results['containerValidity'] = {
    'yamlBlockCount': len(yaml_blocks),
    'mermaidBlockCount': len(mermaid_blocks),
    'validYaml': False
}

yaml_data = None
if yaml_blocks:
    try:
        yaml_data = yaml.safe_load(yaml_blocks[0])
        results['containerValidity']['validYaml'] = True
    except Exception as e:
        results['containerValidity']['error'] = str(e)

ds_ids = []
duplicates = []
actions = set()

if yaml_data:
    screens = yaml_data.get('screens', [])
    routes = yaml_data.get('routes', [])
    metadata = yaml_data.get('metadata', {})
    satisfies_count = len(metadata.get('satisfies', []))
    
    def collect_from_element(el):
        if 'ds_id' in el:
            if el['ds_id'] in ds_ids:
                duplicates.append(el['ds_id'])
            ds_ids.append(el['ds_id'])
        if 'actions' in el:
            for a in el['actions']:
                if isinstance(a, dict):
                    actions.add(a.get('id', a.get('action_id', a.get('name'))))
                else:
                    actions.add(a)
        if 'children' in el:
            for child in el['children']:
                collect_from_element(child)
        if 'states' in el:
            states = el['states']
            if isinstance(states, dict):
                states = states.values()
            for st in states:
                if 'actions' in st:
                    for a in st['actions']:
                        if isinstance(a, dict):
                            actions.add(a.get('id', a.get('action_id', a.get('name'))))
                        else:
                            actions.add(a)
                if 'children' in st:
                    for child in st['children']:
                        collect_from_element(child)
                        
    for screen in screens:
        collect_from_element(screen)
        
    results['prdCoverage'] = {
        'satisfiesCount': satisfies_count,
        'screensCount': len(screens),
        'routesCount': len(routes)
    }
    
    results['componentTraceability'] = {
        'uniqueDsIdsCount': len(set(ds_ids)),
        'duplicates': duplicates
    }

mermaid_events = set()
if mermaid_blocks:
    lines = mermaid_blocks[0].split('\n')
    for line in lines:
        match = re.search(r'-->.*?:\s*(.*)', line)
        if match:
            mermaid_events.add(match.group(1).strip())
            
unmatched_actions = [a for a in actions if a not in mermaid_events]
unmatched_events = [e for e in mermaid_events if e not in actions]

results['logicCoverage'] = {
    'actionsCount': len(actions),
    'eventsCount': len(mermaid_events),
    'unmatchedActions': unmatched_actions,
    'unmatchedEvents': unmatched_events
}

def check_json(filename):
    p = os.path.join(dir_path, filename)
    if not os.path.exists(p):
        return {'exists': False}
    try:
        with open(p, 'r', encoding='utf-8') as f:
            data = json.load(f)
        return {'exists': True, 'valid': True, 'count': len(data)}
    except Exception:
        return {'exists': True, 'valid': False}

results['derivedArtifacts'] = {
    'storyboards': check_json('storyboards.json'),
    'componentMap': check_json('component-map.json'),
    'layoutRules': check_json('layout-rules.json'),
    'artifactIndex': check_json('artifact-index.json')
}

results['derivedArtifacts']['contextSlicesExists'] = os.path.exists(os.path.join(dir_path, 'context-slices'))
results['derivedArtifacts']['reviewDiagramsExists'] = os.path.exists(os.path.join(dir_path, 'review-diagrams.md'))

results['conflictsPreview'] = {
    'previewManifestExists': os.path.exists(os.path.join(dir_path, 'preview', 'preview-manifest.json')),
    'conflictsExists': os.path.exists(os.path.join(dir_path, 'prd-ds-conflicts.md'))
}

print(json.dumps(results, indent=2))
